net/net_test.go

```diff
@@ -21,6 +21,268 @@ import (
+			expectedResult: false,
+			expectError:    true,
+		},
+		{
+			ips:            []string{"1.1.1.1", "2.2.2.2", "3.3.3.3"},
+			errMessage:     "should fail because all are v4",
+			expectedResult: false,
+			expectError:    false,
+		},
+		{
+			ips:            []string{"1.1.1.1", "1.1.1.1"},
+			errMessage:     "should fail because ips are duplicate",
+			expectedResult: false,
+			expectError:    true,
+		},
+
+		{
+			ips:            []string{"fd92:20ba:ca:34f7:ffff:ffff:ffff:ffff", "fd92:20ba:ca:34f7:ffff:ffff:ffff:fff0", "fd92:20ba:ca:34f7:ffff:ffff:ffff:fff1"},
+			errMessage:     "should fail because all are v6",
+			expectedResult: false,
+			expectError:    false,
+		},
+		{
+			ips:            []string{"1.1.1.1", "not-a-valid-ip"},
+			errMessage:     "should fail because 2nd ip is invalid",
+			expectedResult: false,
+			expectError:    true,
+		},
+		{
+			ips:            []string{"not-a-valid-ip", "fd92:20ba:ca:34f7:ffff:ffff:ffff:ffff"},
+			errMessage:     "should fail because 1st ip is invalid",
+			expectedResult: false,
+			expectError:    true,
+		},
+		{
+			ips:            []string{"1.1.1.1", "fd92:20ba:ca:34f7:ffff:ffff:ffff:ffff"},
+			errMessage:     "expected success, but found failure",
+			expectedResult: true,
+			expectError:    false,
+		},
+		{
+			ips:            []string{"fd92:20ba:ca:34f7:ffff:ffff:ffff:ffff", "1.1.1.1", "fd92:20ba:ca:34f7:ffff:ffff:ffff:fff0"},
+			errMessage:     "expected success, but found failure",
+			expectedResult: true,
+			expectError:    false,
+		},
+		{
+			ips:            []string{"1.1.1.1", "fd92:20ba:ca:34f7:ffff:ffff:ffff:ffff", "10.0.0.0"},
+			errMessage:     "expected success, but found failure",
+			expectedResult: true,
+			expectError:    false,
+		},
+		{
+			ips:            []string{"fd92:20ba:ca:34f7:ffff:ffff:ffff:ffff", "1.1.1.1"},
+			errMessage:     "expected success, but found failure",
+			expectedResult: true,
+			expectError:    false,
+		},
+	}
+	// for each test case, test the regular func and the string func
+	for _, tc := range testCases {
+		dualStack, err := IsDualStackIPsStrings(tc.ips)
+		if err == nil && tc.expectError {
+			t.Errorf("%s", tc.errMessage)
+			continue
+		}
+		if err != nil && !tc.expectError {
+			t.Errorf("failed to run test case for %v, error: %v", tc.ips, err)
+			continue
+		}
+		if dualStack != tc.expectedResult {
+			t.Errorf("%v for %v", tc.errMessage, tc.ips)
+		}
+	}
+
+	for _, tc := range testCases {
+		ips := make([]net.IP, 0, len(tc.ips))
+		for _, ip := range tc.ips {
+			parsedIP := net.ParseIP(ip)
+			ips = append(ips, parsedIP)
+		}
+		dualStack, err := IsDualStackIPs(ips)
+		if err == nil && tc.expectError {
+			t.Errorf("%s", tc.errMessage)
+			continue
+		}
+		if err != nil && !tc.expectError {
+			t.Errorf("failed to run test case for %v, error: %v", tc.ips, err)
+			continue
+		}
+		if dualStack != tc.expectedResult {
+			t.Errorf("%v for %v", tc.errMessage, tc.ips)
```

Comment:
good form to print the index, too, e.g `[%d]` prefix

Reply:
we are printing the entire value that generated the error..

Reply:
But that doesn't help me fins which testcase it was.  Minor point, I just have debugged enough of these where the error string didn't actually show me which case of the 30 cases was failing.  NOt super critical unless making other changes

Reply:
Better form is to use [`t.Run`](https://golang.org/pkg/testing/#T.Run)

Reply:
^ agreed, `errMessage` should be the test name passed into `t.Run`